server/src/fstar_connection.ts

```diff
@@ -0,0 +1,434 @@
+import {
+	URI
+} from 'vscode-uri';
+
+import {
+	Position
+} from 'vscode-languageserver/node';
+
+import { setTimeout } from 'timers/promises';
+
+import { FStar, FStarConfig } from './fstar';
+import { isProtocolInfo, IdeProgress, ProtocolInfo,  IdeSymbol, IdeAutoCompleteResponses, FullBufferQuery, FStarRange, LookupQuery, VfsAdd, AutocompleteRequest, CancelRequest,  AnyIdeQueryResponse, IdeQueryResponseR } from './fstar_messages';
+
+// For full-buffer queries, F* chunks the buffer into fragments and responds
+// with several messages, one for each fragment until the first failing
+// fragment. The stream of messages ends with a full-buffer-finished. This
+// behavior allows for displaying incremental progress while the rest of the
+// buffer is being checked. `StreamedResult` is the type of these partially
+// completed queries.
+export type StreamedResult<T> = Promise<[T, StreamedResult<T> | undefined]>;
+
+export class FStarConnection {
+	private last_query_id: number;
+	// TODO(klinvill): Should we have a stronger type for resolve and reject
+	// here that is restricted to a response type and error type?
+	//
+	// Maps query-ids to promises that will be resolved with the appropriate
+	// response.
+	private pending_responses: Map<number, {resolve: (v: any) => void, reject: (e: any) => void, is_stream: boolean}>;
+	private fstar: FStar;
+	debug: boolean;
+
+	constructor(fstar: FStar, debug: boolean) {
+		this.debug = debug;
```

Comment:
Neat! I'm torn between the more verbose representation I originally had (since it visually lays out the debug field next to the other fields) and the more concise syntax proposed here. I went ahead with the concise version.